Source/include/Ably/ARTChannelOptions.h

```diff
@@ -8,6 +8,7 @@ NS_ASSUME_NONNULL_BEGIN
 /**
  * Passes additional properties to an `ARTRestChannel` object, such as encryption.
  */
+NS_SWIFT_SENDABLE
```

Comment:
Also, I’d suggest taking a look at all of the types that you’ve marked as `Sendable`, to check whether their public API exposes any of these non-sendable types. If so, we’ll have to figure out what to do about that.

Reply:
https://github.com/ably/ably-cocoa/pull/1963/commits/401dd03d23bd0835d6b4f63b7c595af8d9ee9116

Reply:
>Also, I’d suggest taking a look at all of the types that you’ve marked as Sendable, to check whether their public API exposes any of these non-sendable types. If so, we’ll have to figure out what to do about that.

Sure, for example `ARTRest` has `createWithOptions:` where `ARTClientOptions` are not sendable. I'm not sure what we can do about that. Do you have any ideas?

Maybe mark those functions and interfaces as unavailable in swift and add new ones specific for swift concurrency. But this can be a lot to change.

Reply:
The issue that we want to avoid is for somebody to be able to use a type that’s marked as sendable as a way to share a non-sendable value between isolation domains. This can happen if the sendable type, for example, exposes a property which is a non-sendable type.

```swift
class SomethingMarkedAsSendable: @unchecked Sendable /* this is the same as writing NS_SWIFT_SENDABLE in Objective-C */ {
    var nonSendableThing: SomethingNonSendable

    init(somethingNonSendable: SomethingNonSendable) {
        self.nonSendableThing = somethingNonSendable
    }
}

class SomethingNonSendable {
    var counter = 0
}

@MainActor
func useThisSupposedlySendableThing(_ somethingMarkedAsSendable: SomethingMarkedAsSendable) {
    // We’re mutating nonSendableThing.counter on the main thread…
    somethingMarkedAsSendable.nonSendableThing.counter += 1

    Task.detached {
        // and also on the global concurrent executor; this is wrong, 
        // but the compiler allows it because it sees that `somethingMarkedAsSendable` 
        // is sendable and hence allows it to be shared across isolation domains; 
        // the fact that internally it contains something non-sendable is irrelevant
        somethingMarkedAsSendable.nonSendableThing.counter += 1
    }

    somethingMarkedAsSendable.nonSendableThing.counter += 1
}
```

So, what we want to avoid is something like the above, for example where:

- `SomethingNonSendable` is one of our mutable options types
- `SomethingMarkedAsSendable` is a client type that exposes its options as a property

> Sure, for example ARTRest has `createWithOptions:` where `ARTClientOptions` are not sendable. I'm not sure what we can do about that. Do you have any ideas?

That alone isn't a problem; it's only a problem if it enables the user to somehow _mutate_ the non-sendable type across different isolation domains (as in the hypothetical example I gave above).

Reply:
> > I don't think that some of the types here are actually Sendable — e.g. these mutable options types that don't synchronise access to their state. So we shouldn't mark those ones.
> 
> [401dd03](https://github.com/ably/ably-cocoa/commit/401dd03d23bd0835d6b4f63b7c595af8d9ee9116)

You seem to have removed `NS_SWIFT_SENDABLE` from immutable types too e.g. `ARTConnectionDetails`; why?

Reply:
>The issue that we want to avoid is for somebody to be able to use a type that’s marked as sendable

Thanks for explanation 👍

>You seem to have removed NS_SWIFT_SENDABLE from immutable types too e.g. ARTConnectionDetails; why?

Messed up 😆 fixed in https://github.com/ably/ably-cocoa/pull/1963/commits/9048f946b9fb470e8d351e10437695532c429461

Reply:
Here is another one - `ARTRestChannel` exposes `options` as `readonly`, but because it's an `NSObject` you can change its embed property `cipher` which is not `readonly`. Changing any of that is a breaking change. What should we do in such a case?

Reply:
Could you please list all such cases that you've found, then we can think about what to do about them?

Reply:
For the case that you've mentioned, it would be a user error for them to attempt to mutate the channel options via the channel, since we say in the README:

> Objects passed to the library must not be mutated afterwards. They can be safely passed again, or read from; they won't be written to by the library.

But, if we wanted to protect against this user error, one option would be to "freeze" the mutable object once it's been passed to the library — that is, for `ARTChannelOptions` and `ARTCipherParams` to have an `isFrozen` property which the library sets to `YES` upon using it to initialize a channel. And then, we override the setters of these classes to throw an exception if somebody tries to mutate a frozen object (i.e. the object which is exposed via the channel’s `options` property).

Reply:
It's also `ARTRealtimeChannel`.`options`.`params` and `ARTLocalDevice`.`secret` + with all `ARTDeviceDetails` properties. Have no idea why they declared as mutable in the public interface.

Reply:
> ARTRealtimeChannel.options.params

`modes`, as well. I think that the freeze + throw-exception approach that I mentioned above would work for these too.

> ARTLocalDevice.secret + with all ARTDeviceDetails properties

To use the library normally the user is never required to construct one of these objects, are they? If not, I think that we could consider the mutability of these properties to be a bug and mark them all as `readonly`. Maybe users are constructing them for testing purposes, in which case we could provide a memberwise initializer, but we could also just do that if people actually complain.

What do you think?

Reply:
I thought this as well, just mark them as readonly and deal with those who is trying to mutate this object.

>modes, as well. I think that the freeze + throw-exception approach that I mentioned above would work for these too.

Yeah, looks like not much else we can do here.